src/providers/alchemy_provider/provider.rs

```diff
@@ -0,0 +1,252 @@
+use crate::{
+    models::token::{TokenBalance, TokenBalances, TokenMetadata},
+    providers::eth_provider::{
+        contracts::erc20::EthereumErc20,
+        error::EthApiError,
+        provider::{EthProviderResult, EthereumProvider},
+    },
+};
+use async_trait::async_trait;
+use auto_impl::auto_impl;
+use eyre::Result;
+use futures::future::join_all;
+use mongodb::bson::doc;
+use reth_primitives::{Address, BlockId, BlockNumberOrTag, Bytes, B256, U256, U64};
+use reth_rpc_types::{
+    serde_helpers::JsonStorageKey, state::StateOverride, txpool::TxpoolContent, BlockOverrides, FeeHistory, Filter,
+    FilterChanges, Header, Index, RichBlock, SyncStatus, Transaction, TransactionReceipt, TransactionRequest,
+};
+
+#[async_trait]
+#[auto_impl(Arc, &)]
+pub trait AlchemyProvider {
+    /// Retrieves the token balances for a given address.
+    async fn token_balances(
+        &self,
+        address: Address,
+        contract_addresses: Vec<Address>,
+    ) -> EthProviderResult<TokenBalances>;
+    /// Retrieves the metadata for a given token.
+    async fn token_metadata(&self, contract_address: Address) -> EthProviderResult<TokenMetadata>;
+    /// Retrieves the allowance for a given token.
+    async fn token_allowance(
+        &self,
+        contract_address: Address,
+        owner: Address,
+        spender: Address,
+    ) -> EthProviderResult<U256>;
+}
+
+#[derive(Debug, Clone)]
+pub struct AlchemyDataProvider<P: EthereumProvider> {
+    eth_provider: P,
+}
+
+impl<P: EthereumProvider> AlchemyDataProvider<P> {
+    pub const fn new(eth_provider: P) -> Self {
+        Self { eth_provider }
+    }
+}
+
+#[async_trait]
+impl<P: EthereumProvider + Send + Sync + 'static> EthereumProvider for AlchemyDataProvider<P> {
```

Comment:
I still don't understand why we need this @eugypalu

Reply:
I think you are right, I will remove it from all three providers

Reply:
No, sorry, I reviewed the code, and EthereumProvider is used in all three providers:

- in pool_provider to call txpool_content
- in debug to call block_by_hash, etc.
- in alchemy for EthereumErc20

So it seemed like a good solution to add it, but maybe I'm missing something. What do you think?

Reply:
afaiu you need an eth_provider but you don't need to implement the provider trait. I will pull the PR and check locally

Reply:
I pushed a commit, you can check it, this should work

Reply:
I saw the commit, now I understand what you meant. Thanks